[Core] Enable Memory Tiering for vLLM

PanJason opened the pull request:
This is the follow-up PR or #7697 
Goal: Enable memory tiering for vLLM where the blocks evicted from HBM will be swapped to DRAM, and further storage media.

Some issues that have been tackled since RFC:
- The oooxxxooo problem: 1. Need to implement swap out for all cases including prefill, decode and swap-in as long as there are CPU blocks 2. Add timestamp for CPU blocks
- Compatible interface with BlockManagerV1, BlockManagerV2 and Embedded Block Allocator.

Support status now:
- BlockManagerV1 w/o chunked prefill: tested
- BlockManagerV2: unsupported yet

A rough road map:

- [x] Layer-by-layer transferring
- [x] Add storage interface
- [ ] BlockManagerV2

Right now chunked prefill does not support prefix caching yet so we don't test it either.

Also, the code needs to be cleaned and rebased on the newest commit, which I have not done yet.

If BlockManagerV2 is not very urgent, we can drop it for now.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

PanJason committed:
First draft and fix format

PanJason committed:
Fix an oooxxxooo problem

zachzzc commented on vllm/core/block_manager_v1.py:
Can you briefly describe what this function does  ?

zachzzc reviewed (commented):
Left some questions to help me better understand the implementation

PanJason committed:
Refactor. Make adjust_swap optional

PanJason committed:
Refactor

PanJason committed:
Add fake layered transfer

Right now only the potential interface for layered transfer is added.
Because vLLM only supports one cuda stream the transmission and
computation is not really overlapped. Also I am uncertain whether this
way of passing the parameters is good
WIP: Multiple cuda streams

PanJason committed:
Fix bug for disk. Now work for tp=1 pp=1 case

PanJason committed:
Now disk work with tp=n

We have to seperate the stateful allocation and stateless
transmission for the vllm implementation (worker multiple processes)

TKONIY committed:
[Core] Initial support for context caching (#2)

- Add initial support for context caching:
    1. Support the endpoint
    2. Introduce another sequence type is_fixed. Now that is_fixed is also considered as is_finished
    3. Note that now the context caching always resides in HBM because the blocks are marked as allocated, and by default, the allocated blocks will not be swapped to any secondary storage.

TKONIY committed:
Fix context caching online serving

PanJason committed:
Add memory tiering for CC (#3)

* Add example disk swap config. Add unit tests for CC with memory tiering

* Layered transfer for DRAM. Transfer in cuda streams

* Fix the missing arg

* Fix context caching online serving

This commit enables layered transmission for DRAM first. Now the
transmission is done in different cuda streams. xformers, flash infer
and flash attention are supported. Optimized transfer for disk
is still pending.

Cherry-pick Yangshen's commit

---------

Co-authored-by: yangshen <yangshen.d@outlook.com>
